com.unity.shadergraph/Editor/Drawing/PreviewManager.cs

```diff
@@ -243,276 +282,443 @@ public bool HandleGraphChanges()
+                    var scale = m_Graph.previewData.scale;
+                    previewTransform *= Matrix4x4.Scale(scale * Vector3.one * (Vector3.one).magnitude / mesh.bounds.size.magnitude);
+                    previewTransform *= Matrix4x4.Translate(-mesh.bounds.center);
 
-                RenderPreview(masterRenderData, mesh, previewTransform);
-            }
+                    RenderPreview(masterRenderData, mesh, previewTransform);
+                }
 
-            m_SceneResources.light0.enabled = false;
-            m_SceneResources.light1.enabled = false;
+                m_SceneResources.light0.enabled = false;
+                m_SceneResources.light1.enabled = false;
 
-            foreach (var renderData in m_RenderList2D)
-                renderData.NotifyPreviewChanged();
-            foreach (var renderData in m_RenderList3D)
-                renderData.NotifyPreviewChanged();
-            if (renderMasterPreview)
-                masterRenderData.NotifyPreviewChanged();
+                foreach (var renderData in renderList2D)
+                    renderData.NotifyPreviewChanged();
+                foreach (var renderData in renderList3D)
+                    renderData.NotifyPreviewChanged();
+                if (renderMasterPreview)
+                    masterRenderData.NotifyPreviewChanged();
 
-            m_RenderList2D.Clear();
-            m_RenderList3D.Clear();
-            m_NodesToDraw.Clear();
+                m_NodesToDraw.Clear();
+            }
         }
 
         public void ForceShaderUpdate()
         {
             foreach (var data in m_RenderDatas.Values)
             {
-                m_NodesToUpdate.Add(data.shaderData.node);
+                m_NodesNeedsRecompile.Add(data.shaderData.node);
             }
         }
 
-        void UpdateShaders()
+        private static readonly ProfilerMarker ProcessCompletedShaderCompilationsMarker = new ProfilerMarker("ProcessCompletedShaderCompilations");
+        void ProcessCompletedShaderCompilations()
         {
             // Check for shaders that finished compiling and set them to redraw
-            foreach (var renderData in m_RenderDatas.Values)
+            using (ProcessCompletedShaderCompilationsMarker.Auto())
+            using (var nodesCompiled = PooledHashSet<AbstractMaterialNode>.Get())
             {
-                if (renderData.shaderData.isCompiling)
+                foreach (var node in m_NodesCompiling)
                 {
-                    var isCompiled = true;
+                    PreviewRenderData renderData = GetPreviewRenderData(node);
+                    PreviewShaderData shaderData = renderData.shaderData;
+                    Assert.IsTrue(shaderData.passesCompiling > 0);
+
+                    if (shaderData.passesCompiling != renderData.shaderData.mat.passCount)
+                    {
+                        Debug.LogWarning("WARNING, shader pass count changed while compiling preview shader, preview may not complete");
```

Comment:
This warning is useful for you but does it help users at all? It's not actionable. Let's figure out the most useful thing to do in this case.

Reply:
I don't think this warning will ever get hit anymore, but if it does, the master preview will never be displayed.
Having this warning will make it much easier to debug what is happening if the master preview never shows up...

Reply:
I had taken this out completely as it was never getting hit, but added it back in the cleanup; if there's some edge case path that de-initializes SRP while we're compiling, it could still potentially get hit.

Reply:
In that case, what can we tell the user to fix it? Close and re-open the graph? Restart Unity? User-facing errors/warnings should be actionable. We can put this kind of detail in the editor.log so if they submit a bug _we_ can track it down.